kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/AxonSerializers.kt

```diff
@@ -0,0 +1,333 @@
+import kotlinx.serialization.SerializationException
+import kotlinx.serialization.builtins.MapSerializer
+import kotlinx.serialization.builtins.SetSerializer
+import kotlinx.serialization.builtins.nullable
+import kotlinx.serialization.builtins.serializer
+import kotlinx.serialization.descriptors.buildClassSerialDescriptor
+import kotlinx.serialization.descriptors.element
+import kotlinx.serialization.encoding.CompositeDecoder
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.encoding.decodeStructure
+import kotlinx.serialization.encoding.encodeStructure
+import kotlinx.serialization.modules.SerializersModule
+import kotlinx.serialization.modules.polymorphic
+import kotlinx.serialization.modules.subclass
+import org.axonframework.eventhandling.GapAwareTrackingToken
+import org.axonframework.eventhandling.GlobalSequenceTrackingToken
+import org.axonframework.eventhandling.MergedTrackingToken
+import org.axonframework.eventhandling.MultiSourceTrackingToken
+import org.axonframework.eventhandling.ReplayToken
+import org.axonframework.eventhandling.TrackingToken
+import org.axonframework.eventhandling.scheduling.ScheduleToken
+import org.axonframework.eventhandling.scheduling.java.SimpleScheduleToken
+import org.axonframework.eventhandling.scheduling.quartz.QuartzScheduleToken
+import org.axonframework.eventhandling.tokenstore.ConfigToken
+import org.axonframework.messaging.responsetypes.InstanceResponseType
+import org.axonframework.messaging.responsetypes.MultipleInstancesResponseType
+import org.axonframework.messaging.responsetypes.OptionalResponseType
+import org.axonframework.messaging.responsetypes.ResponseType
+import kotlin.reflect.KClass
+
+private val trackingTokenSerializer = PolymorphicSerializer(TrackingToken::class).nullable
+
+val AxonSerializersModule = SerializersModule {
+    contextual(ConfigToken::class) { ConfigTokenSerializer }
+    contextual(GapAwareTrackingToken::class) { GapAwareTrackingTokenSerializer }
+    contextual(MultiSourceTrackingToken::class) { MultiSourceTrackingTokenSerializer }
+    contextual(MergedTrackingToken::class) { MergedTrackingTokenSerializer }
+    contextual(ReplayToken::class) { ReplayTokenSerializer }
+    contextual(GlobalSequenceTrackingToken::class) { GlobalSequenceTrackingTokenSerializer }
+    polymorphic(TrackingToken::class) {
+        subclass(ConfigTokenSerializer)
+        subclass(GapAwareTrackingTokenSerializer)
+        subclass(MultiSourceTrackingTokenSerializer)
+        subclass(MergedTrackingTokenSerializer)
+        subclass(ReplayTokenSerializer)
+        subclass(GlobalSequenceTrackingTokenSerializer)
+    }
+
+    contextual(SimpleScheduleToken::class) { SimpleScheduleTokenSerializer }
+    contextual(QuartzScheduleToken::class) { QuartzScheduleTokenSerializer }
+    polymorphic(ScheduleToken::class) {
+        subclass(SimpleScheduleTokenSerializer)
+        subclass(QuartzScheduleTokenSerializer)
```

Comment:
Does this mean when developers use JobRunr or DbScheduler for deadlines/sheduled events they should add something of their own?

Reply:
Yes, they would have to write their own serializer and add it to a `SerializersModule`.
Note that you can combine multiple `SerializerModule`s into a single one that contains all needed serializers. 

The main reason for this is because `kotlinx.serialization` does not use any runtime mechanic such as reflection for serializing classes. Instead the Kotlin compiler generates dedicated serializers for each (data) class annotated with `@Serializable`. For non-Kotlin or external classes this means that you have to write a dedicated serializer and explicitly add it to a `SerializersModule`. You can then add this module to a `SerialFormat` such as `Json` or `Cbor`.